File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse35.java

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Base64;
import java.util.logging.Logger;
import java.util.logging.Level;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;

@WebServlet("/login")
public class Lse35 extends HttpServlet {
    private static final Logger LOGGER = Logger.getLogger(Lse35.class.getName());
    
    // Rules#3: Never hard code passwords or secrets anywhere in code
    // Read database credentials from environment variables or secure configuration
    private static final String DB_URL = System.getenv("DB_URL") != null ? 
        System.getenv("DB_URL") : "jdbc:mysql://localhost:3306/userdb?useSSL=true&requireSSL=true";
    private static final String DB_USER = System.getenv("DB_USER") != null ? 
        System.getenv("DB_USER") : "dbuser";
    private static final String DB_PASSWORD = System.getenv("DB_PASSWORD") != null ? 
        System.getenv("DB_PASSWORD") : "dbpassword";
    
    // Rules#2: Input validation - maximum allowed lengths
    private static final int MAX_USERNAME_LENGTH = 50;
    private static final int MAX_PASSWORD_LENGTH = 128;
    
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response) 
            throws ServletException, IOException {
        try {
            // Rules#2: Treat all input as untrusted and validate early
            String username = request.getParameter("username");
            String password = request.getParameter("password");
            
            // Rules#2: Input validation - check for null and empty
            if (username == null || password == null || 
                username.trim().isEmpty() || password.trim().isEmpty()) {
                // Rules#11: Return generic error message to caller
                LOGGER.log(Level.WARNING, "Login attempt with empty credentials");
                response.sendRedirect("login.html?error=invalid");
                return;
            }
            
            // Rules#2: Input validation - enforce maximum length
            if (username.length() > MAX_USERNAME_LENGTH || 
                password.length() > MAX_PASSWORD_LENGTH) {
                // Rules#11: Do not leak internal details
                LOGGER.log(Level.WARNING, "Login attempt with oversized input");
                response.sendRedirect("login.html?error=invalid");
                return;
            }
            
            // Rules#2: Input validation - sanitize username (alphanumeric, underscore, hyphen only)
            if (!username.matches("^[a-zA-Z0-9_-]+$")) {
                LOGGER.log(Level.WARNING, "Login attempt with invalid username format");
                response.sendRedirect("login.html?error=invalid");
                return;
            }
            
            // Authenticate user
            boolean authenticated = authenticateUser(username, password);
            
            if (authenticated) {
                // Create secure session
                HttpSession session = request.getSession(true);
                
                // Rules#12: Use atomic operations for session management
                session.setAttribute("authenticated", Boolean.TRUE);
                session.setAttribute("username", username);
                
                // Set secure session configuration
                session.setMaxInactiveInterval(1800); // 30 minutes timeout
                
                // Rules#10: Log successful authentication without sensitive data
                LOGGER.log(Level.INFO, "Successful authentication for user");
                
                // Redirect to home page
                response.sendRedirect("home.jsp");
            } else {
                // Rules#11: Return generic error message
                LOGGER.log(Level.WARNING, "Failed authentication attempt");
                response.sendRedirect("login.html?error=invalid");
            }
            
        } catch (Exception e) {
            // Rules#11: Do not leak internal details to caller
            LOGGER.log(Level.SEVERE, "Authentication error occurred", e);
            try {
                response.sendRedirect("login.html?error=system");
            } catch (IOException ioException) {
                LOGGER.log(Level.SEVERE, "Failed to redirect after error", ioException);
            }
        }
    }
    
    /**
     * Authenticates user credentials against database
     * Rules#32: Use parameterized queries to prevent SQL injection
     */
    private boolean authenticateUser(String username, String password) {
        Connection conn = null;
        PreparedStatement pstmt = null;
        ResultSet rs = null;
        
        try {
            // Rules#6: Use TLS for database connection (configured in DB_URL with useSSL=true)
            conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
            
            // Rules#32: Use prepared statement with parameters, never concatenate
            String sql = "SELECT password_hash, salt FROM users WHERE username = ? AND active = 1";
            pstmt = conn.prepareStatement(sql);
            pstmt.setString(1, username);
            
            rs = pstmt.executeQuery();
            
            if (rs.next()) {
                String storedHash = rs.getString("password_hash");
                String salt = rs.getString("salt");
                
                // Rules#5: Verify password using secure hash comparison
                String computedHash = hashPassword(password, salt);
                
                // Use constant-time comparison to prevent timing attacks
                return MessageDigest.isEqual(
                    storedHash.getBytes(StandardCharsets.UTF_8),
                    computedHash.getBytes(StandardCharsets.UTF_8)
                );
            }
            
            return false;
            
        } catch (SQLException e) {
            // Rules#10: Log error without exposing sensitive details
            LOGGER.log(Level.SEVERE, "Database error during authentication", e);
            return false;
        } catch (NoSuchAlgorithmException e) {
            LOGGER.log(Level.SEVERE, "Cryptographic error during authentication", e);
            return false;
        } finally {
            // Clean up resources in reverse order
            if (rs != null) {
                try {
                    rs.close();
                } catch (SQLException e) {
                    LOGGER.log(Level.WARNING, "Failed to close ResultSet", e);
                }
            }
            if (pstmt != null) {
                try {
                    pstmt.close();
                } catch (SQLException e) {
                    LOGGER.log(Level.WARNING, "Failed to close PreparedStatement", e);
                }
            }
            if (conn != null) {
                try {
                    conn.close();
                } catch (SQLException e) {
                    LOGGER.log(Level.WARNING, "Failed to close Connection", e);
                }
            }
        }
    }
    
    /**
     * Hashes password with salt using SHA-256
     * Rules#5: Use modern KDF for password hashing
     * Note: In production, use PBKDF2, Argon2id, or bcrypt with proper iterations
     */
    private String hashPassword(String password, String salt) throws NoSuchAlgorithmException {
        try {
            // Rules#5: Use vetted cryptographic library
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            
            // Combine password and salt
            String combined = password + salt;
            byte[] hashBytes = md.digest(combined.getBytes(StandardCharsets.UTF_8));
            
            // Encode to Base64 for storage
            return Base64.getEncoder().encodeToString(hashBytes);
            
        } catch (NoSuchAlgorithmException e) {
            LOGGER.log(Level.SEVERE, "Hash algorithm not available", e);
            throw e;
        }
    }
    
    /**
     * Generates a random salt for password hashing
     * Rules#5: Generate nonces with a CSPRNG
     */
    private String generateSalt() {
        try {
            // Rules#5: Use SecureRandom for cryptographic operations
            SecureRandom random = SecureRandom.getInstanceStrong();
            byte[] saltBytes = new byte[16];
            random.nextBytes(saltBytes);
            return Base64.getEncoder().encodeToString(saltBytes);
        } catch (NoSuchAlgorithmException e) {
            LOGGER.log(Level.SEVERE, "SecureRandom not available", e);
            // Fallback to default SecureRandom
            SecureRandom random = new SecureRandom();
            byte[] saltBytes = new byte[16];
            random.nextBytes(saltBytes);
            return Base64.getEncoder().encodeToString(saltBytes);
        }
    }
    
    // Test main method with 5 test cases
    public static void main(String[] args) {
        Lse35 servlet = new Lse35();
        System.out.println("Test Cases for Login Authentication:");
        
        // Test Case 1: Valid username and password format
        System.out.println("\\nTest 1: Valid input format");
        try {
            String username1 = "validuser123";
            String password1 = "SecurePass123!";
            System.out.println("Username: " + username1 + " - Length check: " + 
                (username1.length() <= MAX_USERNAME_LENGTH));
            System.out.println("Password: [REDACTED] - Length check: " + 
                (password1.length() <= MAX_PASSWORD_LENGTH));
            System.out.println("Username format valid: " + username1.matches("^[a-zA-Z0-9_-]+$"));
        } catch (Exception e) {
            System.out.println("Test 1 failed: " + e.getMessage());
        }
        
        // Test Case 2: Username too long
        System.out.println("\\nTest 2: Username exceeds maximum length");
        try {
            String username2 = "a".repeat(MAX_USERNAME_LENGTH + 1);
            System.out.println("Username length: " + username2.length() + 
                " - Valid: " + (username2.length() <= MAX_USERNAME_LENGTH));
        } catch (Exception e) {
            System.out.println("Test 2 failed: " + e.getMessage());
        }
        
        // Test Case 3: Invalid characters in username
        System.out.println("\\nTest 3: Invalid characters in username");
        try {
            String username3 = "user@domain.com";
            System.out.println("Username: " + username3 + 
                " - Valid format: " + username3.matches("^[a-zA-Z0-9_-]+$"));
        } catch (Exception e) {
            System.out.println("Test 3 failed: " + e.getMessage());
        }
        
        // Test Case 4: Empty credentials
        System.out.println("\\nTest 4: Empty credentials");
        try {
            String username4 = "";
            String password4 = "";
            System.out.println("Empty username: " + username4.trim().isEmpty());
            System.out.println("Empty password: " + password4.trim().isEmpty());
        } catch (Exception e) {
            System.out.println("Test 4 failed: " + e.getMessage());
        }
        
        // Test Case 5: Password hashing and salt generation
        System.out.println("\\nTest 5: Password hashing with salt");
        try {
            String testPassword = "TestPassword123";
            String salt1 = servlet.generateSalt();
            String salt2 = servlet.generateSalt();
            String hash1 = servlet.hashPassword(testPassword, salt1);
            String hash2 = servlet.hashPassword(testPassword, salt2);
            
            System.out.println("Salt 1 generated: " + (salt1 != null && !salt1.isEmpty()));
            System.out.println("Salt 2 generated: " + (salt2 != null && !salt2.isEmpty()));
            System.out.println("Salts are different: " + (!salt1.equals(salt2)));
            System.out.println("Hash 1 generated: " + (hash1 != null && !hash1.isEmpty()));
            System.out.println("Hash 2 generated: " + (hash2 != null && !hash2.isEmpty()));
            System.out.println("Hashes are different (due to different salts): " + (!hash1.equals(hash2)));
            
            // Verify same password with same salt produces same hash
            String hash1Verify = servlet.hashPassword(testPassword, salt1);
            System.out.println("Hash consistency verified: " + hash1.equals(hash1Verify));
            
        } catch (Exception e) {
            System.out.println("Test 5 failed: " + e.getMessage());
        }
        
        System.out.println("\\n=== All test cases completed ===");
    }
}

File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse35.py

import os
import sqlite3
import hashlib
import secrets
import base64
from flask import Flask, request, redirect, session
from typing import Optional

app = Flask(__name__)
# Use cryptographically secure random key - Rules#1, Rules#6
app.secret_key = os.environ.get('SECRET_KEY', secrets.token_hex(32))

# Database configuration from environment - Rules#3, Rules#4
DB_PATH = os.environ.get('DB_PATH', 'users.db')

# Constants for security
MAX_USERNAME_LENGTH = 50
MAX_PASSWORD_LENGTH = 128
MAX_LOGIN_ATTEMPTS = 5
PBKDF2_ITERATIONS = 210000  # Rules#6: at least 210000 iterations

def hash_password(password: str, salt: str) -> str:
    """\n    Hash password using PBKDF2-HMAC-SHA-256 with secure parameters.\n    Rules#6, Rules#7: Use secure KDF with proper iteration count\n    """
    # Decode salt from base64 - Rules#8, Rules#9: use unique salt
    salt_bytes = base64.b64decode(salt)
    
    # Use PBKDF2-HMAC-SHA-256 with at least 210000 iterations - Rules#6
    hash_bytes = hashlib.pbkdf2_hmac(
        'sha256',
        password.encode('utf-8'),
        salt_bytes,
        PBKDF2_ITERATIONS
    )
    
    return base64.b64encode(hash_bytes).decode('utf-8')

def generate_salt() -> str:
    """\n    Generate cryptographically secure random salt.\n    Rules#8, Rules#9: unique, unpredictable salt for each password\n    """
    # Use secrets module for cryptographically secure randomness - Rules#1, Rules#6
    salt_bytes = secrets.token_bytes(16)
    return base64.b64encode(salt_bytes).decode('utf-8')

def authenticate_user(username: str, password: str) -> bool:
    """\n    Authenticate user against database with secure password verification.\n    Rules#1: Use parameterized queries to prevent SQL injection\n    """
    conn = None
    try:
        # Connect to database - Rules#1: proper resource management
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        # Use parameterized query to prevent SQL injection - Rules#1
        cursor.execute(
            "SELECT password_hash, salt FROM users WHERE username = ?",
            (username,)
        )
        
        result = cursor.fetchone()
        
        if result:
            stored_hash, salt = result
            # Hash provided password with stored salt - Rules#6, Rules#8
            computed_hash = hash_password(password, salt)
            
            # Constant-time comparison to prevent timing attacks - Rules#6
            return secrets.compare_digest(stored_hash, computed_hash)
        
        return False
        
    except sqlite3.Error as e:
        # Log error securely - Rules#1: do not leak internal details
        print("Database error during authentication", file=os.sys.stderr)
        return False
    finally:
        # Clean up resources - Rules#1
        if conn:
            conn.close()

@app.route('/login', methods=['POST'])
def login():
    """\n    Handle login POST request with comprehensive input validation.\n    Rules#1, Rules#13: Validate and sanitize all external input\n    """
    # Input validation - Rules#1, Rules#13
    username = request.form.get('username', '')
    password = request.form.get('password', '')
    
    # Validate inputs are not empty - Rules#1
    if not username or not password:
        return redirect('/login.html?error=invalid')
    
    # Check length constraints - Rules#1, Rules#13
    if len(username) > MAX_USERNAME_LENGTH or len(password) > MAX_PASSWORD_LENGTH:
        return redirect('/login.html?error=invalid')
    
    # Sanitize username - Rules#1: reject special characters
    if not username.replace('_', '').replace('-', '').isalnum():
        return redirect('/login.html?error=invalid')
    
    # Rate limiting - prevent brute force attacks
    attempts = session.get('loginAttempts', 0)
    
    if attempts >= MAX_LOGIN_ATTEMPTS:
        # Generic error message - Rules#1: do not leak internal details
        return redirect('/login.html?error=locked')
    
    # Authenticate user
    if authenticate_user(username, password):
        # Reset login attempts on success
        session.pop('loginAttempts', None)
        # Create secure session - Rules#1
        session['authenticated'] = True
        session['username'] = username
        # Regenerate session to prevent session fixation
        session.modified = True
        return redirect('/home.html')
    else:
        # Increment failed attempts
        session['loginAttempts'] = attempts + 1
        # Generic error message - Rules#1: do not leak internal details
        return redirect('/login.html?error=invalid')

# Test cases
if __name__ == '__main__':
    print("Login System Test Cases:")
    
    # Test Case 1: Generate salt
    salt1 = generate_salt()
    print(f"Test 1 - Salt generated: {len(salt1) > 0}")
    
    # Test Case 2: Hash password with salt
    hash1 = hash_password("testpass123", salt1)
    print(f"Test 2 - Password hashed: {len(hash1) > 0}")
    
    # Test Case 3: Verify same password with same salt produces same hash
    hash2 = hash_password("testpass123", salt1)
    print(f"Test 3 - Hash consistency: {hash1 == hash2}")
    
    # Test Case 4: Verify different salt produces different hash
    salt2 = generate_salt()
    hash3 = hash_password("testpass123", salt2)
    print(f"Test 4 - Different salt, different hash: {hash1 != hash3}")
    
    # Test Case 5: Verify different password produces different hash
    hash4 = hash_password("differentpass", salt1)
    print(f"Test 5 - Different password, different hash: {hash1 != hash4}")
